src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -0,0 +1,127 @@
+ * 3. Bucketwise combine smaller files into 1 file per bucket.
+ * 4. Update index snapshot to remove small files and keep large files + newly created files.
+ *
+ * NOTE: This is an index-only operation. It does not look at the current state of
+ * the data at all. If the data was changed after index creation, optimize will NOT
+ * include the data changes.
+ *
+ * Available modes:
+ * `Quick` mode: This mode allows for fast optimization. Files smaller than a
+ * predefined threshold "spark.hyperspace.index.optimize.fileSizeThreshold" will be picked
+ * for compaction.
+ *
+ * `Full` mode: This allows for slow but complete optimization. ALL files are
+ * eligible for compaction. Compaction threshold is inifinity.
+ *
+ * TODO: Optimize can be a no-op if there is only one small file per bucket.
+ *  https://github.com/microsoft/hyperspace/issues/204.
+ *
+ * @param spark SparkSession
+ * @param logManager IndexLogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ * @param mode Acceptable optimize modes are `quick` and `full`.
+ */
+class OptimizeAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager,
+    mode: String)
+    extends RefreshActionBase(spark, logManager, dataManager) {
+  final override def op(): Unit = {
+    // Rewrite index from small files.
+    val numBuckets = previousIndexLogEntry.numBuckets
+    val indexDF = spark.read.parquet(smallFiles.map(_.name): _*)
+
+    val repartitionedDf =
+      indexDF.repartition(numBuckets, indexConfig.indexedColumns.map(indexDF(_)): _*)
+
+    repartitionedDf.write.saveWithBuckets(
+      repartitionedDf,
+      indexDataPath.toString,
+      logEntry.asInstanceOf[IndexLogEntry].numBuckets,
+      indexConfig.indexedColumns)
+  }
+
+  override def validate(): Unit = {
+    super.validate()
+
+    if (!OPTIMIZE_MODES.exists(_.equalsIgnoreCase(mode))) {
+      throw HyperspaceException(s"Unsupported optimize mode '$mode' found.")
+    }
+
+    if (smallFiles.isEmpty) {
+      throw NoChangesException(
+        s"Optimize aborted as no index files smaller than " +
+          s"${HyperspaceConf.optimizeFileSizeThreshold(spark)} found.")
+    }
+  }
+
+  private lazy val (smallFiles, largeFiles): (Seq[FileInfo], Seq[FileInfo]) = {
+    if (mode.equalsIgnoreCase(IndexConstants.OPTIMIZE_MODE_QUICK)) {
+      val threshold = HyperspaceConf.optimizeFileSizeThreshold(spark)
+      previousIndexLogEntry.content.fileInfos.toSeq.partition(_.size < threshold)
+    } else {
+      (previousIndexLogEntry.content.fileInfos.toSeq, Seq())
+    }
+  }
+
+  override def logEntry: LogEntry = {
+    // Use the `previousLogEntry`. Update content of index files replacing `smallFiles` with newly
+    // created files.
+    val absolutePath = PathUtils.makeAbsolute(indexDataPath)
+    val newContent = Content.fromDirectory(absolutePath)
+    if (largeFiles.nonEmpty) {
+      val largeFilesDirectory: Directory = {
```

Comment:
```suggestion
      val largeFilesDirectory = {
```

Reply:
Is this not taken?

Reply:
i must have missed it. I have added it now, thanks.